### PR TITLE
feat: add multipart form handling

### PR DESCRIPTION
**Pull request by dnlup**
- [x] implementation
- [x] tests
- [x] docs 

@mcollina This is a quick and hopefully not too dirty implementation of [what has been discussed about multipart forms](https://github.com/mcollina/autocannon/issues/160#issuecomment-537889310). If this approach is ok I can continue adding tests and docs.

Note:

this handles only multipart/form-data.

### CODE REVIEW

**Comment by mcollina** [comment]
Would you mind:

1. remove package-lock, and add it to gitignore
2. add tests for this feature

Il giorno ven 18 ott 2019 alle 09:58 dnlup <notifications@github.com> ha
scritto:

**Comment by dnlup** [comment]
@mcollina Sorry for the late response. I added tests, docs will follow too.

File: lib/run.js [hunk 1]
```diff
@@ -93,13 +94,25 @@ function _run (opts, cb, tracker) {
...
+    }
+  }
 
   // copy over fields so that the client
   // performs the right HTTP requests
   url.pipelining = opts.pipelining
-  url.method = opts.method
-  url.body = opts.body
-  url.headers = opts.headers
+  url.method = opts.form ? 'POST' : opts.method
```
Comment (mcollina) [inline lib/run.js]: I think we should honor the method option even if there is a form

File: test/runMultipart.test.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        })
+      })
+      await t.test('error', t => {
+        t.notEqual(null, err)
+        t.equal(c.message, err.message, `mismatching error message ${err.message}`)
+        t.end()
+      })
+    }))
+  }
+  await Promise.all(tests)
```
Comment (mcollina) [inline test/runMultipart.test.js]: There is no need to await them all, it’s done by tap automatically.

File: lib/multipart.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+const { readFileSync } = require('fs')
+const FormData = require('form-data')
+
+function getFormData (string) {
+  try {
+    return JSON.parse(string)
+  } catch (error) {
+    try {
+      return require(string)
```
Comment (mcollina) [inline lib/multipart.js]: Can you please use fe.readFileSync instead? require is relative to the folder of autocannon, not the current directory.

Comment (dnlup) [reply lib/multipart.js]: Yes. This was my lazy approach to load both `json` and `js` files and I didn't think of that.

Comment (dnlup) [reply lib/multipart.js]: would something like this also solve the issue?
```js
const path = require.resolve(string, { paths: [process.cwd()] })
return require(path)
```

Comment (mcollina) [reply lib/multipart.js]: I would prefer the use of fs.readFile anyway, it shows intent.

Comment (dnlup) [reply lib/multipart.js]: Ok

File: lib/multipart.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+const FormData = require('form-data')
+
+function getFormData (string) {
+  try {
+    return JSON.parse(string)
+  } catch (error) {
+    try {
+      const path = resolve(string)
+      const data = readFileSync(path, 'utf8')
+      return JSON.parse(data)
```
Comment (dnlup) [inline lib/multipart.js]: Here I am not allowing `.js` files anymore since it was being a little trickier than plain `.json`.

File: test/runMultipart.test.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+      message: 'A \'type\' key with value \'text\' or \'file\' should be specified'
+    }
+  ]
+
+  for (const c of cases) {
+    t.test(c.name, async t => {
+      const [err] = await new Promise((resolve) => {
+        run({
+          url: 'http://localhost:' + server.address().port,
+          method: 'POST',
```
Comment (dnlup) [inline test/runMultipart.test.js]: Now that the `opts.method` in honored I am passing the method in the test calls.

Comment (mcollina) [reply test/runMultipart.test.js]: Ok!

File: lib/multipart.js (hunk 4 above)
Comment (mcollina) [reply lib/multipart.js]: that’s ok!

File: lib/run.js (hunk 1 above)
Comment (dnlup) [reply lib/run.js]: I have restored
```js
url.method = opts.method
```

but I am not sure that is what you meant anymore.

Maybe I could have assigned a default and let the user override it?

```js
if (opts.form) {
  opts.method = opts.method || 'POST'
}
```

Comment (mcollina) [reply lib/run.js]: Yes, exactly.

**Comment by dnlup** [comment]
Will do!

File: help.txt [hunk 6]
```diff
@@ -47,7 +51,7 @@ Available options:
   -D/--reconnectRate NUM
         Some number of requests to make before resetting a connections connection to the
         server.
-  -n/--no-progress
+  -n/--no-progressallowed.includes(req.method)
```
Comment (mcollina) [inline help.txt]: Is this change deliberate?

Comment (dnlup) [reply help.txt]: No, sorry I'll fix it.

**Comment by dnlup** [comment]
Sorry for the latest force pushes. I realized I committed with my work email address and GitHub was not recognizing me.

File: test/runMultipart.test.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+      name: 'missing value in text type',
+      value: '{ "image": { "type": "text" }}',
+      message: 'Missing key \'value\' in form object for key \'image\''
+    },
+    {
+      name: 'missing value in text type',
+      value: '{ "image": { "type": "random" }}',
+      message: 'A \'type\' key with value \'text\' or \'file\' should be specified'
+    }
+  ]
```
Comment (mcollina) [inline test/runMultipart.test.js]: I think we should support an API where the form value is a JS object, and not just the json version of it

File: lib/run.js [hunk 8]
```diff
@@ -93,13 +97,25 @@ function _run (opts, cb, tracker) {
...
 
   // copy over fields so that the client
   // performs the right HTTP requests
   url.pipelining = opts.pipelining
   url.method = opts.method
-  url.body = opts.body
-  url.headers = opts.headers
+  url.body = form ? form.getBuffer() : opts.body
+  url.headers = form ? {
+    ...Object.assign({}, opts.headers, form.getHeaders())
```
Comment (mcollina) [inline lib/run.js]: there is no need for the object spread here

File: test/runMultipart.test.js [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+      })
+      await t.test('error', t => {
+        t.notEqual(null, err)
+        t.equal(c.message, err.message, `mismatching error message ${err.message}`)
+        t.end()
+      })
+    })
+  }
+
+  t.end()
```
Comment (mcollina) [inline test/runMultipart.test.js]: I think this t.end is superflous
